tslearn/early_classification.py

```diff
@@ -0,0 +1,269 @@
+                    # to fullfil the criterions of the cost function
+                    np.fill_diagonal(conf_matrix, 0)
+                    self.pyhatyck_[t - self.minimum_time_stamp, k] = conf_matrix
+
+    def get_cluster_probas(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        diffs = X[np.newaxis, :] - self.cluster_.cluster_centers_[:, :len(X)]
+        distances_clusters = np.linalg.norm(diffs, axis=(1, 2))
+        minimum_distance = np.min(distances_clusters)
+        average_distance = np.mean(distances_clusters)
+        minimum_distance_to_average = (
+            average_distance - minimum_distance
+        ) / average_distance
+        delta = (average_distance - distances_clusters) / average_distance
+        probas = 1.0 / (
+            np.exp(self.lamb * minimum_distance_to_average)
+            + np.exp(self.lamb * (minimum_distance_to_average - delta))
+            + 1e-6
+        )
+        return probas / probas.sum()
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[0]
+        sum_pyhatyck = np.sum(self.pyhatyck_[truncated_t + tau - self.minimum_time_stamp], axis=1)
+        sum_pyhatyck = np.transpose(sum_pyhatyck)
+        sum_global = np.sum((sum_pyhatyck * self.pyck_), axis=0)
+        cost = np.dot(proba_clusters, sum_global)
+        return cost + self._cost_time(truncated_t + tau)
+
+    def cost_function_minimizer(self, end_of_time, xt):
+        """
+        We want to minimize a function "funct" according a time series "xt" for a list of integers from 0 to
+        "end_of_time"
+
+        Parameters
+        ----------
+        end_of_time: int
+            the highest integer at which the function to be minimized is calculated
+        function_to_minimize: funct
+            the function of interest
+        xt: vector
+            the time series related to the function "funct"
+
+        Returns
+         ------
+         int : The integer in {0,...,stop} that minimizes "funct"
+         float : the so fat minimum cost
+        """
+        costs = [self._expected_cost(xt, tau) for tau in range(end_of_time + 1)]
```

Comment:
In fact, this seems to be the only place where `_expected_cost` is called. This means we need expected costs for all future timestamps, so probably a future step will be to have `_expected_cost` take a single input argument and return an array of expected costs (and vectorize it :). Can be kept for later, clearly, but it would make sense in my opinion.

Reply:
Also, it seems that `end_of_time` argument for `cost_function_minimizer` would no longer be required if this code was vectorized.